File: python/fluent_cards/builders/refresh_builder.py
from __future__ import annotations
from typing import Callable


class RefreshBuilder:
    def __init__(self):
        self._refresh: dict = {}

    def with_action(self, configure: Callable) -> RefreshBuilder:
        from .action_builder import ActionBuilder
        b = ActionBuilder()
        configure(b)
        self._refresh['action'] = b.build()
        return self

    def add_user_id(self, user_id: str) -> RefreshBuilder:
        if 'userIds' not in self._refresh:
            self._refresh['userIds'] = []
        self._refresh['userIds'].append(user_id)
        return self

    def with_expires(self, expires: str) -> RefreshBuilder:
        self._refresh['expires'] = expires
        return self

    def build(self) -> dict:
        return self._refresh

File: node/packages/fluent-cards/src/builders/RefreshBuilder.ts
import type { RefreshConfiguration } from '../models.js';
import { ActionBuilder } from './ActionBuilder.js';

/** Fluent builder for {@link RefreshConfiguration}. */
export class RefreshBuilder {
  private readonly refresh: RefreshConfiguration = {};

  /** Sets the action used to refresh the card. @param configure A callback to configure the ActionBuilder. @returns The builder instance for method chaining. */
  withAction(configure: (b: ActionBuilder) => void): this {
    const b = new ActionBuilder();
    configure(b);
    this.refresh.action = b.build();
    return this;
  }

  /** Adds a user ID that is permitted to trigger the refresh. @param userId The user ID. @returns The builder instance for method chaining. */
  addUserId(userId: string): this {
    (this.refresh.userIds ??= []).push(userId);
    return this;
  }

  /** Sets the expiry timestamp for the refresh. @param expires The ISO 8601 expiry timestamp. @returns The builder instance for method chaining. */
  withExpires(expires: string): this {
    this.refresh.expires = expires;
    return this;
  }

  /** Builds and returns the configured RefreshConfiguration. @returns The configured RefreshConfiguration instance. */
  build(): RefreshConfiguration { return this.refresh; }
}
